File: apps/backend/app/repositories/media_progress_repository.py
from app.repositories.base_repository import BaseRepository
from app.models.media_progress import MediaProgress
from app.models.media import Media
from app.models.libraries import Library
from sqlalchemy import select
from sqlalchemy.dialects.postgresql import insert

from uuid import UUID

class MediaProgressRepository(BaseRepository[MediaProgress]):
    model = MediaProgress
    allowed_updates = {'current_position', 'last_watched', 'last_device_id'}

    async def fetch_media_progress_validate(
        self, user_id, media_id
    ) -> MediaProgress | None:
        """user -> lib -> media -> media progress"""
        stmt = (
            select(self.model)
            .join(Media, MediaProgress.media_id == Media.id)
            .join(Library, Media.library_id == Library.id)
            .where(
                Library.user_id == user_id,
                Media.id == media_id,
            )
        )

        result = await self.db.execute(stmt)

        return result.scalar_one_or_none()

    async def fetch_all_by_user(self, user_id: UUID) -> list[MediaProgress]:

        query = (
            select(self.model)
            .join(Media, self.model.media_id == Media.id)
            .join(Library, Media.library_id == Library.id)
            .where(Library.user_id == user_id)
        )
        res = await self.db.execute(query)

        return list(res.scalars().all())

    async def save_or_update(self, media_progress: MediaProgress):

        query = insert(MediaProgress).values(
            media_id=media_progress.media_id,
            current_position=media_progress.current_position,
            last_device_id=media_progress.last_device_id,
        )

        query = query.on_conflict_do_update(
            index_elements=[MediaProgress.media_id],
            set_={
                "current_position": query.excluded.current_position,
                "last_device_id": query.excluded.last_device_id,
            },
        )
        await self.db.execute(query)

    async def is_media_owned_by_user(self, media_id: UUID, user_id: UUID) -> bool:
        query = (
            select(Media.id)
            .join(Library, Library.id == Media.library_id)
            .where(Media.id == media_id, Library.user_id == user_id)
        )

        res = await self.db.execute(query)
        return res.scalar_one_or_none() is not None

    async def update_media_progress_validate(self, entity_id: UUID, user_id: UUID, **kwargs) -> bool:

        permitted_progress = (
            select(self.model.id)
            .join(Media, self.model.media_id == Media.id)
            .join(Library, Media.library_id == Library.id)
            .where(self.model.id == entity_id, Library.user_id == user_id)
        )

        return await self.update_entity_validate(permissions=permitted_progress, **kwargs)

File: apps/frontend/lib/db/repositories/media_progress_repository.ts
import { MediaProgress } from "@/lib/types/MediaProgress";
import { SQLiteDBConnection } from "@capacitor-community/sqlite";
import { persistDatabase } from "../database";
import { EntityType } from "@/lib/types/EntityType";

type MediaProgressRow = {
  id: string;
  media_id: string;
  current_position: number | null;
  last_watched: string;
  last_device_id: string | null;
  version: number;
};

export class MediaProgressRepository {
  private db: SQLiteDBConnection;

  constructor(dbConnection: SQLiteDBConnection) {
    this.db = dbConnection;
  }
  async initTable(): Promise<void> {
    await this.db.execute(`
        CREATE TABLE IF NOT EXISTS media_progresses (
            id TEXT PRIMARY KEY NOT NULL,

            media_id TEXT NOT NULL,
            current_position TEXT,
            last_watched TEXT NOT NULL,
            last_device_id TEXT,

            version INTEGER NOT NULL DEFAULT 0,

            FOREIGN KEY (media_id)
            REFERENCES media(id)
            ON DELETE CASCADE,

            UNIQUE (media_id)
        );
        `);
  }

  async upsertFromSync(progress: MediaProgress): Promise<void> {
    await this.db.run(
      `
    INSERT INTO media_progresses (
      id,
      media_id,
      current_position,
      last_watched,
      last_device_id,
      version
    )
    VALUES (?, ?, ?, ?, ?, ?)

    ON CONFLICT(media_id) DO UPDATE SET
      id = excluded.id,
      current_position = excluded.current_position,
      last_watched = excluded.last_watched,
      last_device_id = excluded.last_device_id,
      version = excluded.version;
    `,
      [
        progress.id,
        progress.mediaId,
        progress.currentPosition,
        progress.lastWatched,
        progress.lastDeviceId,
        progress.version,
      ],
    );

    await persistDatabase();
  }

  async getMediaProgressById(
    userId: string,
    libraryId: string,
    mediaId: string,
  ): Promise<MediaProgress | null> {
    const res = await this.db.query(
      `SELECT mp.* FROM media_progresses mp
    JOIN media m ON mp.media_id = m.id
    JOIN libraries l ON m.library_id = l.id WHERE l.user_id = ?
    AND m.library_id = ?
    AND mp.media_id = ?
    `,
      [userId, libraryId, mediaId],
    );

    const row = res.values?.[0];
    if(!row)
      {
        return null;
      }

    return this.mapRowToObject(row);
  }
  async saveMediaProgress(progress: MediaProgress): Promise<void> {
    await this.db.run(
      `
    INSERT INTO media_progresses (
      id,
      media_id,
      current_position,
      last_watched,
      last_device_id,
      version
    )
    VALUES (?, ?, ?, ?, ?, ?)
    `,
      [
        progress.id,
        progress.mediaId,
        progress.currentPosition,
        progress.lastWatched,
        progress.lastDeviceId,
        progress.version,
      ],
    );

    await persistDatabase();
  }

  async upsertMediaProgress(progress: MediaProgress): Promise<void> {
    await this.db.run(
      `
    INSERT INTO media_progresses (
      id,
      media_id,
      current_position,
      last_watched,
      last_device_id,
      version
    )
    VALUES (?, ?, ?, ?, ?, ?)

    ON CONFLICT(media_id) DO UPDATE SET
      current_position = excluded.current_position,
      last_watched = excluded.last_watched,
      last_device_id = excluded.last_device_id,
      version = excluded.version
    `,
      [
        progress.id,
        progress.mediaId,
        progress.currentPosition,
        progress.lastWatched,
        progress.lastDeviceId,
        progress.version,
      ],
    );

    await persistDatabase();
  }
  mapRowToObject(row: MediaProgressRow): MediaProgress {
    return {
      id: row.id,
      mediaId: row.media_id,
      currentPosition: row.current_position,
      lastWatched: row.last_watched,
      lastDeviceId: row.last_device_id,
      version: row.version,
      entityType: EntityType.MediaProgress,
    } as MediaProgress;
  }
}
